pkg/dockerregistry/server/util.go

```diff
@@ -0,0 +1,83 @@
+package server
+
+import (
+	"fmt"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+const (
+	// EventReasonPushDenied is a name of reason for denial of blob write or manifest put. May be used by
+	// event listeners to filter the events on image stream objects.
+	EventReasonPushDenied = "PushDenied"
+)
+
+// autoProvisionImageStream attempts to create an image stream for given repository in etcd. An existence
+// check must done before calling this function.
+func autoProvisionImageStream(ctx context.Context, repo *repository) (*imageapi.ImageStream, error) {
+	stream := &imageapi.ImageStream{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: repo.name,
+		},
+	}
+
+	client, ok := UserClientFrom(ctx)
+	if !ok {
+		return nil, fmt.Errorf("Error creating user client to auto provision image stream: Origin user client unavailable")
+	}
+
+	stream, err := client.ImageStreams(repo.namespace).Create(stream)
+	if err != nil {
+		statusErr, ok := err.(*kerrors.StatusError)
+		if !ok {
+			return nil, fmt.Errorf("Error auto provisioning image stream: %v", err)
+		}
+		if kerrors.IsAlreadyExists(statusErr) {
+			context.GetLogger(ctx).Infof("Cannot create image stream %s/%s that already exists", repo.namespace, repo.name)
+			return client.ImageStreams(repo.namespace).Get(repo.name)
+		}
+		if kerrors.IsForbidden(statusErr) {
+			context.GetLogger(ctx).Errorf("Denied auto provisioning image stream: %v", statusErr)
+			return nil, distribution.ErrAccessDenied
+		}
+		return nil, fmt.Errorf("Error auto provisioning image stream: %v", statusErr)
+	}
+
+	return stream, nil
+}
+
+// reportPushDeniedEvent reports an event on image stream coresponding to the given repository about a
+// failed push due to quota abuse. The is argument must be fetched from api server so that its selfLink is
+// filled. If nil, the image stream will be created. Message string will be formatted with given arguments.
+func reportPushDeniedEvent(ctx context.Context, repo *repository, is *imageapi.ImageStream, message string, args ...interface{}) {
```

Comment:
Because the error description doesn't get propageted to the http response. Upstream wants it that way. Without the event intercepted by the builder, all the build logs will offer is `requested access to the resource is denied`.

Creating an event seemed to me as a better option compared to storing the error on image stream itself.